addons/mail/models/mail_channel.py

```diff
@@ -724,30 +726,51 @@ def _execute_channel_pin(self, pinned=False):
         if channel_partners:
             channel_partners.write({'is_pinned': pinned})
 
-    def channel_seen(self):
+    def channel_seen(self, last_message_id=None):
+        """
+        Mark channel as seen by updating seen message id of the current logged partner
+        :param last_message_id: the id of the message to be marked as seen, last message of the
+        thread by default. This param SHOULD be required, the default behaviour is DEPRECATED and
+        kept only for compatibility reasons.
+        """
         self.ensure_one()
-        if self.channel_message_ids.ids:
-            last_message_id = self.channel_message_ids.ids[0] # zero is the index of the last message
-            channel_partner = self.env['mail.channel.partner'].search([('channel_id', 'in', self.ids), ('partner_id', '=', self.env.user.partner_id.id)], limit=1)
-            if channel_partner.seen_message_id.id == last_message_id:
-                # last message seen by user is already up-to-date
+        domain = [('channel_ids', 'in', self.ids)]
+        if last_message_id:
+            domain = expression.AND([domain, [('id', '<=', last_message_id)]])
+        last_message = self.env['mail.message'].search(domain, order="id DESC", limit=1)
+        if not last_message:
+            return
+
+        channel_partner = self.env['mail.channel.partner'].search([
+            ('channel_id', 'in', self.ids),
+            ('partner_id', '=', self.env.user.partner_id.id)
+        ], limit=1)
+        if channel_partner:
+            if channel_partner.seen_message_id and channel_partner.seen_message_id.id >= last_message.id:
```

Comment:
Why did you need to revert using query here? It should have been fine, if there was an issue with it maybe removing it is hiding another problem.

Reply:
The problem is that there is an edge case that does not work with the query : a partner that have no channel partner for the given channel. In that case, you'll never find a channel partner in the query but now because the conditions on seen_message_id are not satisfied but just because the partner didn't see any messages in this channel and no channel partner has been created for him. (Spotted that with the channel_fetch python test)

Reply:
Users that are not members of the channel should not be able to do channel seen. If there is a test trying to do that, it should be adapted instead.

Reply:
Okay, wasn't sure about that :)
Nevertheless, I got a problem with `=?` operator :
```python
channel_partner_domain = expression.AND([
            [('channel_id', 'in', self.ids)],
            [('partner_id', '=', self.env.user.partner_id.id)],
            expression.OR([
                [('seen_message_id', '=?', False)],
                [('seen_message_id', '<', last_message.id)]
            ])
        ])
        channel_partner = self.env['mail.channel.partner'].search(channel_partner_domain, limit=1)
```

If I have a channel partner with a last seen message id of 265 and my last_message.id is 266, I still get the channel partner (and I don't really know why) ... So I consider the following change :

```python
channel_partner_domain = expression.AND([
            [('channel_id', 'in', self.ids)],
            [('partner_id', '=', self.env.user.partner_id.id)],
            expression.OR([
                [('seen_message_id', '=', False)],
                [('seen_message_id', '=', None)],
                [('seen_message_id', '<', last_message.id)]
            ])
        ])
        channel_partner = self.env['mail.channel.partner'].search(channel_partner_domain, limit=1)
```

What do you think ?

Reply:
Yes I just noticed `=?` is not made to work in db... There's a specific condition in `expression.py` that turns it into `TRUE` in that case.

I guess I misunderstood the operator as checking for `NULL` or `false` in db, whereas it actually just checks for `None` or `False` in its right operator in the domain.

You can actually drop either line and go back to initial vertsion. What I mentioned appears to be only important for boolean fields actually. `expression.py` has that for non-boolean fields both (False or None) are going to be converted to `IS NULL` is db, which is what we want.

Relational fields should never contain actually `false` in db, and the query builder is smart enough for that case even if you give `False` as param.